### ISSUE TITLE
`grad2F1` can terminate too early when only some gradients are requested

### ISSUE THREAD

**Issue by ricardoV94** [stan-dev/math#2857]
The termination criteria uses the maximum change for the 3 gradients (wrt a1, a2 and b1), which makes it less strict when only a subset is being evaluated.

https://github.com/stan-dev/math/blob/95abd90d38259f27c7a6013610fbc7348f2fab4b/stan/math/prim/fun/grad_2F1.hpp#L117

I get an absolute difference of ~`4.3e-4` across the two evaluations shown in the snippet:
```cpp
#include <stan/math.hpp>
#include <iostream>

int main() {

  double a1 = -0.5;
  double a2 = -4.5;
  double b1 = 11.0;
  double z = 0.3;

  auto grad_tuple1 = stan::math::internal::grad_2F1_impl<true, true, true, true>(a1, a2, b1, z);
  std::cout << std::get<0>(grad_tuple1) << std::endl;

  auto grad_tuple2 = stan::math::internal::grad_2F1_impl<true, false, false, false>(a1, a2, b1, z);
  std::cout << std::get<0>(grad_tuple2) << std::endl;
}
```
```bash
# Number of iterations:9 
-0.122702
# Number of iterations:2
-0.122727
```

The result obtained in the second case would probably fail the precision tested here: https://github.com/stan-dev/math/blob/2e174d22b8ad9d9fd4c6a1e0aeff48700b74c951/test/unit/math/prim/fun/grad_2F1_test.cpp#L116

Perhaps a simple solution is to introduce a minimum number of iterations?

**Comment by andrjohns** [comment]
Thanks so much for catching this (I was the one that introduced this behaviour)! I've opened a PR to add a minimum number of iterations

**Comment by andrjohns** [comment]
@ricardoV94 Just a heads up that I've also found a couple of minor bugs in the `grad_2F1` implementation while testing this, and will have a PR up next week with the bugfixes and more comprehensive tests

**Comment by ricardoV94** [comment]
@andrjohns thanks for the heads up. Where those bugs restricted to the Euler approximation by any chance?

**Comment by andrjohns** [comment]
There was one bug with the Euler approximation on L213](https://github.com/stan-dev/math/blob/3cbf8e1d978a780d03b4e465fcc0fcd8ae53bde9/stan/math/prim/fun/grad_2F1.hpp#L213):
```cpp
auto hyper_da2 = hypergeometric_2F1(a1_euler, a2, b1, z_euler);
```

The second argument should be `a2_euler`:
```cpp
auto hyper_da2 = hypergeometric_2F1(a1_euler, a2_euler, b1, z_euler);
```

The other bugs were template/type-deduction issues, so probably not relevant outside of Stan & c++

There's also an issue with using higher-order autodiff when `z = 0` (some entries of the Hessian matrix are incorrect), but the first-order derivatives are all fine - still chasing that one down.

**Comment by andrjohns** [comment]
Closed by #2858
